docs/rfcs/20220413-aot-for-all-snode.md

```diff
@@ -0,0 +1,340 @@
+
+# Goals
+
+* Provide a SNode API that allows SNodes and Taichi fields to be localized, so that Taichi kernels can be made *pure*.
+* Supports describing the complete SNode tree type explicitly.
+* Make SNode types serializable into AOT, so that AOT can use all kinds of SNodes.
+* The new SNode API should offer compatibility with the existing usage.
+* (Uncertain, but highly desired) Decouple the element type from the SNode type, i.e. the situation where matrix fields has to been implemented in the "scattered" way to support SoA layout.
+
+## Non-Goals
+
+* Expand the support for sparse SNodes beyond LLVM's codegen, especially SPIR-V.
+
+# Detailed Design
+
+## A first attempt
+
+This API would clearly allow us to pass in fields into kernels:
+
+```py
+a = ti.field(ti.i32)
+b = ti.field(ti.f32)
+ti.root.pointer(ti.ij, 16).dense(ti.ij, 16).place(a, b)
+
+@ti.kernel
+def run(a: ?, b: ?):
+  for I in ti.grouped(a):
+    b[I] = a[I] * 4.2
+
+run(a, b)
+```
+
+However, it doesn't really work for AOT, because `a` and `b` are **attributes of a tree type**. That is, you *cannot* dump `a` and `b`'s types separately.
+
+To understand this problem, we can define something equivalent in C++:
+
+```cpp
+struct AB {
+  int32_t a;
+  float b;
+};
+
+using TreeType = PointerDense<AB>;
+```
+
+We cannot declare the `run` kernel as `void run(? a, ? b)`. Instead, we have to pass in a `TreeType` instance as a whole into the kernel, i.e., `void run(TreeType &tree)`.
+
+Internally, as you are using Taichi's SNode system to construct hierarchies, you are also constructing a `SNodeTree` type at the same time. This is done by Taichi's [`FieldsBuilder`](https://github.com/taichi-dev/taichi/blob/master/python/taichi/_snode/fields_builder.py).
+
+## A working design
+
+We will make the SNode tree and its type explicit by providing `SNodeTreeBuilder`. Each field needs to be registered into the builder via `add_field()`. `add_field()` does *not* actually do any memory allocation. Instead, it just returns a *field ID*, which can be used to retrieve a field from the tree inside the kernel.
+
+```py
+builder = ti.SNodeTreeBuilder()
+
+x = builder.add_field(dtype=ti.f32, name='x')
+y = builder.add_field(dtype=ti.i32, name='y')
+builder.tree()
+  .pointer(ti.ij, 4)
+  .dense(ti.ij, 5)
+  .place(x, y)
+
+# `tree_t` stands for "tree type".
+tree_t = builder.build()
+```
+
+Similarly, `SNodeTreeBuilder.build()` doesn't allocate memory for the tree. It only builds *the type of* a SNode tree. You can later instantiate a tree with `tree_t.instantiate()`. There are a few reasons behind this type-tree decoupling design:
+
+1. We have explicit access to the SNode tree type. This is a must for AOT, but can also be used as type annotations for enhanced language formality.
+2. We can instantiate as many trees as we want from this type.
```

Comment:
-- and pass them to the same kernel without re-compilation.

This seems to allow SNodes to offer similar functionalities to ndarrays, as it enables the same kernel to operate on different buffers without recompiling. 

How feasible do you think it would be to take this one step further, and allow kernels to operate on different instances of one-layer dense fields without recompilation?

For example, it would be great if we could support
```py
x = builder.add_field(dtype=ti.f32, name='x')
builder.tree()
  .dense(ti.ij, 5)
  .place(x)

tree_t = builder.build()

x_t = tree_t.field_type("x")

@ti.kernel
def f(x: x_t):
  for i in x:
    ....
```
Or, simpler still:
```py
@ti.kernel
def f(x: ti.types.field(ti.f32, (N)) )
  for i in x:
    ...
```

Reply:
This is an interesting thought. I haven't really figured out how to best describe the type for an individual field. Your suggestion does remind me to make one-layer dense field and Ndarrays interchangeable, i.e. one is the view of the other.